docs/architecture/ADR-005-liquidation.md

```diff
@@ -0,0 +1,191 @@
+
+### Basic Message Type
+
+To implement the liquidation functionality of the Asset Facility, one message type is required:
+
+```go
+// MsgLiquidate - a liquidator targets a specific borrower, asset type, and collateral type for complete or partial liquidation
+type MsgLiquidate struct {
+  Liquidator    sdk.AccAddress
+  Borrower      sdk.AccAddress
+  Repayment     sdk.Coin // denom + amount
+  RewardDenom   string
+}
+```
+
+Repayment's denom is the borrowed asset denom to be repaid (because the borrower may have multiple open borrows). It is always a base asset type (not a uToken).
+Its amount is the maximum amount of asset the liquidator is willing to repay. This field enables partial liquidation.
+
+RewardDenom is the collateral type which the liquidator will recieve in exchange for repaying the borrower's loan. It is always a uToken denomination.
+
+It is necessary that messages be signed by the liquidator's account. Thus the method `GetSigners` should return the `Liquidator` address for the message type above.
+
+### Partial Liquidation
+
+It is possible for the amount repaid by a liquidator to be less than the borrower's total borrow position in the target denom. Such partial liquidations can succeed without issue, reducing the borrowed amount by `Repayment.Amount` and rewarding collateral proportional to the amount repaid.
+
+Additionally, the `Repayment.Amount` found in `MsgLiquidate` functions as a maximum the liquidator is willing to repay. There are multiple factors that may reduce the actual repayment below this amount:
+
+- Borrowed amount of `Repayment.Denom` is below `Repayment.Amount`
+- Insufficient `RewardDenom` collateral to match repaid value plus liquidation incentive
+- Parameter-based maximum, e.g. if a parameter `LiquidationCap = 0.50` caps liquidated value in a single transaction at 50% of the total borrowed value.
+
+In the above scenarios, the `MsgLiquidate` should succeed with the maximum amount allowed by the conditions, rather than fail outright.
+
+### Token Parameters
+
+In order to incentivize liquidators to target certain collateral types for liquidation first, the token parameter `LiquidationIncentive` is used.
+
+When a `MsgLiquidate` causes liquidation to occur, the liquidator receives collateral equal to (100% + `RewardDenom.LiquidationIncentive`) of the repaid value worth of collateral.
+
+For example, if the liquidation incentive for `uatom` is `0.15`, then the liquidator receives `u/uatom` collateral worth 115% of the borrowed base assets they repaid. The denom of the base assets does not affect this calculation.
+
+### Calculating Liquidation Amounts
+
+When a `MsgLiquidate` is received, the `x/leverage` module must determine if the targeted borrow address is eligible for liquidation.
+
+```go
+    // from MsgLiquidate (liquidatorAddr, borrowerAddr, repayDenom, repayAmount, rewardDenom)
+
+    borrowed := GetTotalBorrows(borrowerAddr)
```

Comment:
We need to do this per-token right? Since `GetTotalBorrows` returns all borrowed tokens.

Reply:
GetTotalBorrows returns an `sdk.Coins` with all borrowed token types. We sum the value of all those coins when computing total borrowed value (so if we only build oracle with a single-denom `GetValue`, then we will have to call it once for each denom in the `sdk.Coins`).

Reply:
Ok, that's not super clear from this pseudo-code but I don't think it's a big deal. Feel free to modify it and make it clearer at your own discretion 👍